boa/src/syntax/parser/statement/iteration/for_statement.rs

```diff
@@ -88,42 +88,64 @@ where
             _ => Some(Expression::new(true, self.allow_yield, self.allow_await).parse(cursor)?),
         };
 
+        let can_iterate = match &init {
+            Some(Node::VarDeclList(v)) => v.as_ref().len() == 1,
+            Some(Node::ConstDeclList(v)) => v.as_ref().len() == 1,
+            Some(Node::LetDeclList(v)) => v.as_ref().len() == 1,
+            _ => false
+        };
+
         // TODO: for..in, for..of
         match cursor.peek(0)? {
-            Some(tok) if tok.kind() == &TokenKind::Keyword(Keyword::In) => {
-                unimplemented!("for...in statement")
+            Some(tok) if tok.kind() == &TokenKind::Keyword(Keyword::In) && can_iterate => {
+                cursor.next()?;
+                let variable = init.unwrap();
+                let object = Expression::new(true, self.allow_yield, self.allow_await).parse(cursor)?;
+                cursor.expect(
+                    TokenKind::Punctuator(Punctuator::CloseParen),
+                    "for..in statement",
+                )?;
+                let body =
+                    Statement::new(self.allow_yield, self.allow_await, self.allow_return).parse(cursor)?;
+                Ok(ForInLoop::new(variable, object, body).into())
             }
-            Some(tok) if tok.kind() == &TokenKind::identifier("of") => {
+            Some(tok) if tok.kind() == &TokenKind::identifier("of") && can_iterate => {
```

Comment:
Is `TokenKind::identifier("of")` a placeholder, or intentional? i.e. Should there be `TokenKind::Keyword(Keyword::Of)`?